docs/process/guidance/feature_flags/index.rst

```diff
@@ -0,0 +1,507 @@
+
+    {
+        "some_key": @SOME_FEATURE@
+    }
+
+.. code-block:: bazel
+    :caption: some_module/BUILD
+
+    expand_template(
+        name = "config",
+        out = "config.json",
+        substitutions = select({
+            ":config_some_feature": {
+                "@SOME_FEATURE@": "value_for_some_feature"
+            },
+            "//conditions:default": {
+                "@SOME_FEATURE@": "some_default"
+            },
+        }),
+        template = "config.json.tmpl",
+    )
+
+Testing
+-------
+
+With the support of Path Mapping, we are able to fully take advantage of a
+machine's resources and correctly support an arbitrary set of configurations.
+We can setup Bazel targets / test_suites which set their configuration as code
+and run them all in parallel. Developers are able to run all tests with one
+command while avoiding duplication.
+
+Since Bazel will not support shared actions, the deduplication mechanism in
+this case will be the use of a cache. There are still a couple of minor
+drawbacks with cache deduplication:
+
+- There is a slight inefficiency in a first execution if two equal actions
+  under different configurations start at the same time without any cache
+  entry.
+- Using a disk_cache as a deduplication mechanism implies developers should be
+  aware of this and regularly use caches in their day-to-day business, which is
+  generally recommended anyways.
+
+Note that Path Mapping must be individually supported by every rule.
+
+Unit tests
+^^^^^^^^^^
+
+A unit test shall never test more than one single feature.
+
+The selection which tests shall be executed depending on a feature values shall
+be done via preprocessor macros in the test.cpp
+
+The activation/deactivation of features is propagated via `local_defines`.
+
+.. code-block:: cpp
+    :caption: some_module/test.cpp
+
+    #if defined(SOME_FEATURE)
+        EXPECT_CALL(...)
+    #else
+        EXPECT_CALL(...)
+
+.. code-block:: bazel
+    :caption: some_module/BUILD
+
+    cc_test(
+        name = "test",
+        srcs = ["test.cpp"],
+        deps = [":lib"],
+        local_defines = select({
+            ":config_some_feature": ["SOME_FEATURE"],
+        "//conditions:default": [],
+        }),
+    )
+
+    cc_library(
+        name = "lib",
+        local_defines = select({
+            ":config_some_feature": ["SOME_FEATURE"],
+            "//conditions:default": [],
+        }),
+        # ...
+    )
+
+Test with the default feature value:
+
+.. code-block:: bash
+
+    bazel test :test
+
+Test with the feature value explicitly enabled
+
+.. code-block:: bash
+
+    bazel test :test --//some_module:some_feature=True
+
+Feature flag discovery
+----------------------
+
+Available feature flags can be found with bazel cqueries.
+
+.. code-block:: bash
+
+    bazel cquery \
+      "filter('.*',
+          kind('.*_flag', deps('//TODO Main target to be defined with overall integration'))
+      )" --output label_kind | sort
```

Comment:
Do you expect people to write such cqueries? Then they need a lot more explanation.

Reply:
Ah I missed that this will produce a list of features. So users do not need to modify that cquery. Can we put that into some alias? `baze list-feature-flags` would be nice :-)